packages/terra-disclosure-manager/src/DisclosureManager.jsx

```diff
@@ -0,0 +1,401 @@
+   * actions occur.
+   */
+  app: AppDelegate.propType,
+  /**
+   * The child components that will be provided with an AppDelegate 'app' prop used to interact with the DisclosureManager instance.
+   */
+  children: PropTypes.node,
+  /**
+   * A function used to provide rendering capabilities to the DisclosureManager.
+   */
+  render: PropTypes.func.isRequired,
+  /**
+   * An array of disclosure types that the DisclosureManager should support. If an unsupported disclosure request occurs, the DisclosureManager will
+   * utilize its 'app' prop and forward the request instead of handling the request itself.
+   */
+  supportedDisclosureTypes: PropTypes.array,
+};
+
+const defaultProps = {
+  supportedDisclosureTypes: [],
+};
+
+class DisclosureManager extends React.Component {
+  /**
+   * Clones the current disclosure component state objects and returns the structure for further mutation.
+   */
+  static cloneDisclosureState(state) {
+    const newState = Object.assign({}, state);
+    newState.disclosureComponentKeys = Object.assign([], newState.disclosureComponentKeys);
+    newState.disclosureComponentData = Object.assign({}, newState.disclosureComponentData);
+
+    return newState;
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.resolveDismissPromise = this.resolveDismissPromise.bind(this);
+    this.resolveDismissChecksInSequence = this.resolveDismissChecksInSequence.bind(this);
+
+    this.disclosureTypeIsSupported = this.disclosureTypeIsSupported.bind(this);
+    this.safelyCloseDisclosure = this.safelyCloseDisclosure.bind(this);
+    this.generatePopFunction = this.generatePopFunction.bind(this);
+
+    this.renderContentComponents = this.renderContentComponents.bind(this);
+    this.renderDisclosureComponents = this.renderDisclosureComponents.bind(this);
+
+    this.openDisclosure = this.openDisclosure.bind(this);
+    this.pushDisclosure = this.pushDisclosure.bind(this);
+    this.popDisclosure = this.popDisclosure.bind(this);
+    this.closeDisclosure = this.closeDisclosure.bind(this);
+    this.requestDisclosureFocus = this.requestDisclosureFocus.bind(this);
+    this.releaseDisclosureFocus = this.releaseDisclosureFocus.bind(this);
+    this.maximizeDisclosure = this.maximizeDisclosure.bind(this);
+    this.minimizeDisclosure = this.minimizeDisclosure.bind(this);
+
+    // These cached functions are stored outside of state to prevent unnecessary rerenders.
+    this.dismissChecks = {};
+    this.onDismissResolvers = {};
+
+    this.state = {
+      disclosureIsOpen: false,
+      disclosureIsFocused: true,
+      disclosureIsMaximized: false,
+      disclosureSize: undefined,
+      disclosureComponentKeys: [],
+      disclosureComponentData: {},
+    };
+  }
+
+  disclosureTypeIsSupported(type) {
+    const { app, supportedDisclosureTypes } = this.props;
+
+    return supportedDisclosureTypes.indexOf(type) >= 0 || !app;
+  }
+
+  openDisclosure(data) {
+    this.setState({
```

Comment:
Certainly not something that needs to be addressed with this pr, but should a repeated disclosure event trigger a close (like tapping a row twice)?

Reply:
I think implementing such logic within the DisclosureManager itself would muddy up the APIs a bit. We should find a way to implement and share that behavior elsewhere. 

The only way we'd be able to do that here is by comparing the disclosure keys, which doesn't seem like a foolproof solution.

Reply:
Yeah, hopefully elsewhere isn't everywhere except the DisclosureManager (like in every component implementation). We can figure out that later if need be. Just a thought.